server/ServerFSM.cpp

```diff
@@ -371,36 +380,110 @@ sc::result Idle::react(const Error& msg) {
     return discard_event();
 }
 
+sc::result Idle::react(const Hostless&) {
+    return transit<MPLobby>();
+}
+
 
 ////////////////////////////////////////////////////////////
 // MPLobby
 ////////////////////////////////////////////////////////////
+
+namespace {
+    GG::Clr GetUnusedEmpireColour(const std::list<std::pair<int, PlayerSetupData>>& psd) {
+        //DebugLogger(FSM) << "finding colours for empire of player " << player_name;
+        GG::Clr empire_colour = GG::Clr(192, 192, 192, 255);
+        for (const GG::Clr& possible_colour : EmpireColors()) {
+            //DebugLogger(FSM) << "trying colour " << possible_colour.r << ", " << possible_colour.g << ", " << possible_colour.b;
+
+            // check if any other player / empire is using this colour
+            bool colour_is_new = true;
+            for (const std::pair<int, PlayerSetupData>& entry : psd) {
+                const GG::Clr& player_colour = entry.second.m_empire_color;
+                if (player_colour == possible_colour) {
+                    colour_is_new = false;
+                    break;
+                }
+            }
+
+            // use colour and exit loop if no other empire is using the colour
+            if (colour_is_new) {
+                empire_colour = possible_colour;
+                break;
+            }
+
+            //DebugLogger(FSM) << " ... colour already used.";
+        }
+        return empire_colour;
+    }
+
+    // return true if player has important changes.
+    bool IsPlayerChanged(const PlayerSetupData& lhs, const PlayerSetupData& rhs) {
+        return (lhs.m_client_type != rhs.m_client_type) ||
+            (lhs.m_starting_species_name != rhs.m_starting_species_name) ||
+            (lhs.m_save_game_empire_id != rhs.m_save_game_empire_id) ||
+            (lhs.m_empire_color != rhs.m_empire_color);
+    }
+}
+
 MPLobby::MPLobby(my_context c) :
     my_base(c),
-    m_lobby_data(new MultiplayerLobbyData()),
-    m_server_save_game_data(new ServerSaveGameData())
+    m_lobby_data(new MultiplayerLobbyData(std::move(Server().m_galaxy_setup_data))),
+    m_server_save_game_data(new ServerSaveGameData()),
+    m_ai_count(1)
```

Comment:
this line doesn't seem to do what it's supposed to... I get default MP lobbies with no AIs in them

Reply:
AI players doesn't stored in `GalaxySetupData` so they cann't be reused. Only galaxy options and game rules are saved,

Reply:
Only AI count could be reused from connection information.

Reply:
The AI count is reproduced when going back to the lobby, but the issue is that the initial lobby state has 0 AIs, even though it looks like it's supposed to be 1 AI by default based on this line...

Reply:
`m_ai_count` is supposed to replace static `AI_count` which uses for AI name's postfix. Maybe it better to rename this variable.